pypdf/_font.py

```diff
@@ -28,13 +32,152 @@ class FontDescriptor:
     character_widths: dict[str, int] = field(default_factory=dict)
 
     @classmethod
-    def from_font_resource(cls, pdf_font_dict: DictionaryObject) -> "FontDescriptor":
+    def from_font_resource(
+        cls,
+        pdf_font_dict: DictionaryObject,
+        encoding: Optional[Union[str, dict[int, str]]] = None,
+        char_map: Optional[dict[Any, Any]] = None
+    ) -> "FontDescriptor":
+        from pypdf._cmap import get_encoding  # noqa: PLC0415
         from pypdf._codecs.core_fontmetrics import CORE_FONT_METRICS  # noqa: PLC0415
         # Prioritize information from the PDF font dictionary
         font_name = pdf_font_dict.get("/BaseFont", "Unknown").removeprefix("/")
-        if font_name in CORE_FONT_METRICS:
+
+        font_descriptor_obj: Optional[DictionaryObject] = None
+        if not (encoding and char_map):
+            encoding, char_map = get_encoding(pdf_font_dict)
+
+        if "/FontDescriptor" in pdf_font_dict:
+            font_descriptor_obj = pdf_font_dict.get("/FontDescriptor")
+        elif font_name in CORE_FONT_METRICS:
             return CORE_FONT_METRICS[font_name]
-        return cls(name=font_name)
+
+        # Initialize font_kwargs with defaults.
+        font_kwargs: dict[str, Any] = {
+            "name": font_name,
+            "family": cls.family,
+            "weight": cls.weight,
+            "ascent": cls.ascent,
+            "descent": cls.descent,
+            "cap_height": cls.cap_height,
+            "x_height": cls.x_height,
+            "italic_angle": cls.italic_angle,
+            "flags": cls.flags,
+            "bbox": (-100.0, -200.0, 1000.0, 900.0),
+            "character_widths": {}
+        }
+
+        # TrueType and Type1 fonts have a /Widths array mapping character codes to widths
+        if isinstance(encoding, dict) and "/Widths" in pdf_font_dict:
+            first_char = pdf_font_dict.get("/FirstChar", 0)
+            font_kwargs["character_widths"] = {
+                encoding.get(idx + first_char, chr(idx + first_char)): width
+                for idx, width in enumerate(cast(ArrayObject, pdf_font_dict["/Widths"]))
+            }
+
+        # CID fonts have a /W array mapping character codes to widths stashed in /DescendantFonts
+        if "/DescendantFonts" in pdf_font_dict:
```

Comment:
Are we able to move this whole conditional block into a dedicated method to keep methods short and readable where possible?